### PR TITLE
Implicit downcast involving references

### PR DESCRIPTION
**Pull request by andersfugmann**
Add tests for implict downcast when using references, and fix the test to detect implicit downcasts when using references, which was not previously not detected.

### CODE REVIEW

**Comment by andersfugmann** [comment]
I'm unsure if a change note is required for this PR.

File: cpp/ql/src/Likely Bugs/Conversion/ImplicitDowncastFromBitfield.ql
Comment (rdmarsh2) [inline cpp/ql/src/Likely Bugs/Conversion/ImplicitDowncastFromBitfield.ql:27]: The `and any()` here is unnecessary.

**Comment by jbj** [comment]
LGTM, modulo @rdmarsh2's comment. Including a change note is the right thing to do. It goes in https://github.com/github/codeql/tree/main/cpp/change-notes, and you can see the style and conventions by looking at a few other recent change notes.

File: cpp/ql/src/Likely Bugs/Conversion/ImplicitDowncastFromBitfield.ql
Comment (andersfugmann) [reply cpp/ql/src/Likely Bugs/Conversion/ImplicitDowncastFromBitfield.ql]: Good catch. I have removed the superfluous code. Thanks

**Comment by andersfugmann** [comment]
I've added a change-note. I'm somewhat unsure on what to write in it, so please do not hesitate to comment so I can learn.

File: cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md
Comment (jbj) [inline cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md:3]:
```suggestion
* The query `cpp/implicit-bitfield-downcast` now accounts for C++ reference types, which leads to move true positive results.
```

I think this is how the old Semmle docs team has taught us to write change notes. Notably, queries are referred to by id (or name and id), and TPs are called "true positive results".

Cc @geoffw0 for a second pair of eyes.

**Review by jbj — COMMENTED** [review]
Otherwise LGTM

File: cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md
Comment (intrigus-lgtm) [reply cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md:3]: Typo: move
which leads to **more** true positive results.

Comment (jbj) [reply cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md:3]: Fixed, thanks! Clearly a second pair of eyes was needed :smile:.

Comment (geoffw0) [reply cpp/change-notes/2021-08-24-implicit-downcast-from-bitfield.md:3]: Yes, we refer to queries by query ID: `cpp/implicit-bitfield-downcast`

Or sometimes with the query name and ID: "Implicit downcast from bitfield" (`cpp/implicit-bitfield-downcast`)

But not by file name in formal places such as this.  Query IDs have the advantages of being very easy to search for, and in theory they should never change and there should never be two distinct queries with the same ID (which can happen with query names, e.g. the equivalent query for two different target languages).

Other than that, as with all documentation try to use simple English, avoid idioms and avoid words with apostrophes ("doesn't").  Technical words are generally fine though.  What you have now LGTM.

**Review by geoffw0 — DISMISSED** [review]
LGTM.

**Review by geoffw0 — COMMENTED** [review]
Oh sorry, the change note change needs accepting, then LGTM.

**Review by jbj — APPROVED** [review]

**Pull request merged**
